### PR TITLE
Fix log depth woes

### LINKED ISSUE #6103: Atmosphere is culled irregularly when near/far frustum is changed
image

```
var viewer = new Cesium.Viewer('cesiumContainer');
viewer.camera.frustum.near = 0.15;
viewer.camera.frustum.far = 50000000;
viewer.camera.rotateRight(0.01);
viewer.camera.rotateDown(0.001);
```

Reported on the forum here: https://groups.google.com/forum/#!topic/cesium-dev/Q-_88AaT8sc

### LINKED ISSUE #6573: log depth interpolation not handled correctly for large geometries
Our "big" geometries aren't tessellated enough for log depth to interpolate correctly right now, which causes issues that look kind of like near-plane problems but don't present as a planar cut:

ezgif com-optimize

[On Sandcastle](https://cesiumjs.org/Cesium/Build/Apps/Sandcastle/#c=ZZHLboMwEEV/xWIDkagpaRX6IFHVpLtKldKqKzaOGcCqsaOxIUqq/HsdyLuzmzP3XmvGLUPSClgBkjFRsCJTMKKp6XfHAp937VQry4QC9AfPmcpU76CGgwIqdclQ2KoWfAZLW702RdHFWWzgXA7KCivAUJbnwW+miCvFaiBPxOcV8B/AhWaYEwRumSol+GGvOgIn3Rt3VYEoK+tYTG/DE+ZaYy4Us2DcbL/Q/BBBC9T1DEoEMMFNPBzSOEmSh5Dcj+jdKB49hsTR+EiTng7OHqhdNAomT+lTLTXS+dusF20ztb241Ebr+ksHV4fYSbzQS41dS5j01hdRLzVa0qAMKI0s1Eu52yRaNO5AlnLT2dLoYEpz0RKRjzPv6q8yj3DJjHGTopHyU2wg8yZp5PQXNqmZu1b50QJKtt5Jqnjy3kNKaRq59r/Lai0XDM8S/wA)

I'm pretty sure it's the "minimum tessellation" problem (described here: http://outerra.blogspot.sk/2009/08/logarithmic-z-buffer.html) because adding `#define ENABLE_GL_POSITION_LOG_DEPTH_AT_HEIGHT` to the top of [`vertexLogDepth.glsl`](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Shaders/Builtin/Functions/vertexLogDepth.glsl) seems to fix the problem.

### LINKED ISSUE #7189: Hide atmosphere until globe is ready to render
When Cesium first loads, we often see the white/red/yellow artifacts that are part of the sky atmosphere that are "under/behind" the globe while the globe loads.  It doesn't bother us, but I think it looks like a bug/artifact to a lot of users.  Let's not render it until the globe is "loaded."  This will couple things slightly more but should be easily implemented just in the `Scene` render loop.

@OmarShehata could you look at this one this week with help from @lilleyse or @bagnell?

---

Please close if this is a duplicate; I thought there was an issue but could not find it.

### LINKED ISSUE #8465: PointPrimitive flickers/disappears when disableDepthTestDistance is POSITIVE_INFINITY
I am creating a ```PointPrimitive``` (not an Entity) like so:
```
var point = points.add( {
    position: Cesium.Cartesian3.fromDegrees( -60, 0, 0, viewer.scene.globe.ellipsoid ),
    pixelSize: 20,
    color: new Cesium.Color( 1, 0, 1, 1 ),
    heightReference: Cesium.HeightReference.CLAMP_TO_GROUND,
    disableDepthTestDistance: Number.POSITIVE_INFINITY
} );
```
When the page first loads, the point is sometimes visible, sometimes not. In either case, if I rotate the globe for a few seconds (no particular direction -- just back and forth, up and down), the point flickers badly. Sometimes it disappears entirely, and stays invisible for tens of seconds. Sometimes it seems like the direction I drag matters (e.g. the point stays visible if I drag upward, but disappears if I drag to the left), but that could just be my imagination.

Sandcastle example: [LINK](https://sandcastle.cesium.com/#c=ZVJha9swEP0rwl/iQiY3G+xDloYVp9sMTWIatzAwBNm+JMfOkifJbtOx/z45jkPdCSNLd/fe3b1TIzRrEJ5Bsxsm4ZmFYLAu+dPJ5rNRfrqHSlqBEvRozP6kkrml4XcNxj6ALEAvVQFTZnUN486LpdiDPsZaNej807fcCRIsRbUB3WAO0TDS7/nbVWua9qisRipcnprgUZOr7NYYsCZI4MXWGkywEu4v6E5oe4iiEbvqeP62B7d9SWUqG9dupVBaM2w3bm2xxhItNhAqIsgtKumqYR22E4mbHCTwqo80XBSF31O+y+FSdI4u6NxXpQy21JfGQlewOwn5ie+0Khew1wDGZx8+X49Z9w2S70llwIEIK6OwYFdnySt8AdrgqxvEx+uzLVekhuKHrcVnkxOv2ycX/AFwf3Dz3IEGmcOlvh9DOw/vb5fxNllvvz+sH1eLM7pAIzKCBVT2kLh3sUBjxYlmVZeZKz5eb6IkerrbRqtv0SpKfnZT8cbezNgjwbzj+YplpbRtR+9zHlgoKxJOnyCr819geW5MK/Ms6EGzAhuGxU3qvXuqqcdyEsY4z66mkzKpN58FLn4AIyUKlPt1A5rEsQ05TOb3nZFzPgvc9X+UVYoyod8w/gM)

If I set ```disableDepthTestDistance``` to zero, the problem does not occur. However, in my application that causes other problems -- I have other primitives on the ellipsoid surface, whose vertices stick up a little bit, and the point can end up partially behind those vertices. Disabling the depth test seems like the right way to solve that problem ... but when the depth test is disabled, I encounter the problem described above.

I spent a while digging in the Cesium code, but I can't even figure out the general category of what's going on. The point doesn't seem to be occluded -- sometimes the issue is apparent even before the ellipsoid gets drawn ... nothing is being rendered but the point and the starry background, yet the point still flickers.

Browser: Firefox, Chromium
Operating System: Linux

### LINKED ISSUE #8562: Custom Model rendering issue/default Cesium rendering issue
[ForCesium.zip](https://github.com/AnalyticalGraphicsInc/cesium/files/4099828/ForCesium.zip)

Attached in the Zip file is the model I am using and a video of the issue I am seeing for your reference. 

I work for a company that is using a more detailed model of the one provided, and are seeing this model rendering issue. It is prominent when the model is loaded into a cesium sandbox. However, I also loaded the model into the gltf vs code extension and saw a similar result. The thing that makes me think this is a Cesium problem is from the video you can see I first view the model and animations in Babylon.js and there are no issues, but when I view the model from the Cesium model preview I see the problem. The filament sheet when it is close or intersecting another object with some camera changes there is a blending issue. For instance, if you rotate to the bottom of the model after the animation, the filament is clearly intersecting the black rectangle but with a move of the camera the filament is more prominent and the black rectangle is hidden. Also when the filament is fully animated and you rotate around the purple filament brace blends through the filament.

This is a problem for my employer because the accuracy of the model is of the upmost importance. The modelers of the model have also tried adding UV's and image maps but that did not solve this.

In the sandcastles I have messed around with frustums, camera properties, etc. I was not able to find an immediate solution. 

While you are looking at this issue and determine it is something on the model side that needed to be changed to fix this then let us know what it is so that our modelers can determine how to create for Cesium users, but I would think that cesium out of the box should be able to handle models the way babylon.js or three.js can.

Cesium Version: "1.61.0"
Sandcastle example: Load attached glb or gltf into the 3D models sandcastle to see the blending problem.

Browser: Chrome 79.0.3945.117

Operating System: Windows 10 Enterprise(1809)

Devon Despain

### PR DESCRIPTION
**Pull request by kring**
When log depth is enabled, Cesium has problems with large geometries near the camera getting clipped. #6573 and #7189 are some examples of this.

This is a known problem with the log depth approach when it is implemented using only a modification to the vertex shader. It happens when a large triangle's vertices are outside the frustum, but the triangle itself passes through it. The vertex shader adjusts the clip space Z coordinate to be logarithmic. But then the GPU, in between the vertex and fragment shaders, linearly interpolates the Z coordinates of two vertices to see if they intersect the frustum. Linearly interpolating logarithmic values is not a valid thing to do. It clearly produces incorrect results, and in this case it results in the GPU deciding the triangle does _not_ intersect the frustum. So the triangle generates no fragments, and we see giant ugly holes in our geometry.

It can be "fixed" by making sure geometry is small-ish, so that interpolating logarithmic values linearly doesn't produce such wildly wrong results. That would be challenging in Cesium, and feels a bit error-prone in general.

The other approach is to correctly compute the logarithmic depth values in the fragment shader, rather than relying on its interpolation across triangles. The main downsides to this approach are that it requires the `GL_EXT_frag_depth` extension, and that writing depth in the fragment shader disables early-Z optimizations by the GPU.

But despite the downsides we think this is a win for Cesium versus the alternative (more frustums), so Cesium has since the beginning written log depth in the fragment shader. Not just when it's needed, but always, whenever we're using log depth. So wait, why do we still have problems with large geometry being clipped?

Well, remember I said the large geometry was getting clipped before it ever got to the fragment shader? So, no fragment shader change, including writing log depth, is going to stop that. It's too late. To fix the problem with large geometries, it's not enough to write log depth in the fragment shader. We also have to _not_ write log depth in the vertex shader.

Cesium actually has a mechanism for doing this, by defining `DISABLE_GL_POSITION_LOG_DEPTH` in the vertex shader. Terrain would define this for upsampled tiles and fill tiles, and other parts of Cesium would try to turn it on at various times as well. But anytime we had large geometry and didn't `#define DISABLE_GL_POSITION_LOG_DEPTH`, we'd get geometry clipping problems.

But here's the thing: because we're _always_ writing log depth in the fragment shader, we shouldn't _ever_ need to write log depth in the vertex shader. So why not just have `#define DISABLE_GL_POSITION_LOG_DEPTH` on all the time, always use linear depth in the clip space Z coordinate, and no more large geometry artifacts. Simple!

Except that creates new artifacts when zoomed out:
image

What is that?!

After lots of investigation, I came to the conclusion that it's this:
When using log depth, we use a huge near/far ratio. It's so large that vertices that are nowhere near the far plane are frequently deemed to be on the wrong side of it, because so much error is introduced in the projection transformation and the perspective divide. When this happens, the triangle containing the vertex gets clipped before we get to the vertex shader.

So here's my solution, and you're going to have to hold onto your hat for this one: in this PR, we clamp the clip space Z coordinate to be between the near and far planes. That way we always get fragments, and so we can properly evaluate their depth in the fragment shader. Sounds crazy, right? Let me try to convince you:

* When using log depth, there's pretty much always just one frustum. The tiny amount of geometry that lies in front of the camera but behind the near plane should be irrelevant, and there shouldn't be any geometry past the far plane because we chose the far plane to accommodate all geometry.
* Even Brano Kemen says you're supposed to use depth clamping: https://stackoverflow.com/questions/22364930/logarithmic-depth-buffer-opengl/22374026 (user camenomizoratojoakizunewake).

Now I'm slightly worried about what this will do in a multi-frustum + log depth case, where you ideally wouldn't pass _all_ the geometry overlapping two adjacent frustums through the fragment shader for each frustum. Perhaps we could only clamp depth values that are "reasonably" near the near and far planes to the planes. But that starts to get pretty error prone again. The current approach is reliable, and we can optimize if we can find a scenario where we actually can measure a negative performance impact from it.

So, with these changes, large geometry is always rendered perfectly, and there are no weird z-clipping artifacts like in the image above.

I also spruced up the log depth computation to hopefully make it work better in the (rare?) case that we do have multiple frustums with log depth. Previously, the second frustum wouldn't get as much precision as the first because the log depth values even in the second frustum were measured from a distance of 0, rather than the near plane. In the first frustum those two are basically the same, in the second frustum they're not at all. Actually if I'm thinking about this right, having a second frustum with log depth was totally pointless before. But it's possible I'm not thinking about it right.

I also implemented polygon offset for log depth. This is used in the bivariate depth test for 3D Tiles, and previously worked by turning off fragment shader depth writes (the only place in Cesium that did this!) in order to make fixed function polygonOffset work. But turning off FS depth writes is no longer an option. Please review this part carefully, in writeLogDepth.glsl, cause there's a decent chance I don't know what I'm doing.

Fixes #6573
Fixes #7189
Fixes #6103
Fixes #8465
Fixes #8562

To do before merging:
* [ ] Fix test failures
* [x] Remove / deprecate some old uniforms.
* [x] Make sure 2D and Columbus View work
* [x] Make sure 3D with an orthographic frustum works as well as it ever did.
* [x] Make sure other browsers, GPUs, and operating systems are happy.
* [x] Check if it fixes any of the possiblies above.
* [x] Measure performance. I claim it should be the same or better, but we need to make sure.

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @kring!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/fix-log-depth/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

**Comment by emackey** [comment]
I spent about 20 minutes poking around at this, and I couldn't reproduce any of the logdepth problems that I know of.  This looks really promising!

Although from the screenshot above it looks like maybe there's a rivalry with the West-coast Australians going on? 😄

**Comment by lilleyse** [comment]
I searched github for log depth issues and tried them all out on this branch, but there's no way I got everything. The common theme is polylines still have issues but everything else seems to be fixed. This PR may set the record for most closed issues all at one.

Some of these are also in the PR description, but I just wanted to put them in a single place. Add more as you see them.

https://github.com/AnalyticalGraphicsInc/cesium/issues/6103 (fixed)
https://github.com/AnalyticalGraphicsInc/cesium/issues/6573 (fixed)
https://github.com/AnalyticalGraphicsInc/cesium/issues/6741#issuecomment-406717425 (not fixed)
https://github.com/AnalyticalGraphicsInc/cesium/issues/6911 (need confirmation)
https://github.com/AnalyticalGraphicsInc/cesium/issues/7072 (not fixed)
https://github.com/AnalyticalGraphicsInc/cesium/issues/7074 (curvy artifact is fixed but there's some clipping)
https://github.com/AnalyticalGraphicsInc/cesium/issues/7189 (fixed)
https://github.com/AnalyticalGraphicsInc/cesium/issues/7374 (not fixed)
https://github.com/AnalyticalGraphicsInc/cesium/issues/7360 (polyline doesn't flicker but it gets clipped)
https://github.com/AnalyticalGraphicsInc/cesium/issues/7852 (not fixed)
https://github.com/AnalyticalGraphicsInc/cesium/issues/8465 (tiny bit of flickering, but I consider this fixed)
https://github.com/AnalyticalGraphicsInc/cesium/issues/8562 (fixed)

**Comment by kring** [comment]
Awesome, thanks Sean! I'll take a look at the polyline issues to see if there's a reasonable fix there, too.

**Comment by kring** [comment]
The 7 remaining test failures are all in `Vector3DTilePolygons` and only happen under WebGL2. From what I can tell, the classification primitive isn't being rendered at all, and it's really not clear to me why. After spending hours trying to debug this, I finally realized that the Classification Sandcastle example doesn't work under WebGL2 even in master. See #8629. But these 7 tests _do_ pass in master. 😕 

So I don't know if my log depth changes just make the problem more obvious, or fixed something that previously made the tests just coincidentally pass, or what.

Any chance someone who understands the classification code better could take a quick look?

It would be really great to get this PR into next week's CesiumJS release, and a shame to hold it up with a problem like this.

**Comment by kring** [comment]
I looked at performance using the following Sandcastle examples:
* Terrain
* 3D Tiles Point Cloud Shading
* 3D Tiles BIM
* Clamp to Terrain

In all cases, I used my integrated GPU in Chrome, used the built version of Sandcastle, added the Cesium Inspector mixin, and opened in a new window. I put master on one half of my screen and this branch on the other, making sure the sizes of both were identical. Then I turned on the Inspector performance display and watched the frame rate on each size. They were indistinguishable.

**Comment by mramato** [comment]
@lilleyse what are your thoughts here?  This is definitely an exciting and high impact change, so it would be nice to get it in.  From my perspective, if it's ready to be merged and we do it today, that should be more than enough time to catch any surprises before next release.

**Comment by lilleyse** [comment]
I don't think this needs to be held back because of WebGL 2.

@kring is there a fix on the horizon for polylines? I'm actually still ok with merging before then if all the other artifacts are fixed.

**Comment by kring** [comment]
There's no easy/obvious fix for the polyline problem. I'm still looking at it, and hope to get it fixed eventually, but I think we should get this in in the meantime.

So maybe just exclude the 7 failing WebGL2 tests for now?

**Comment by kring** [comment]
I've disabled the broken WebGL2 tests for now, and added a comment referencing #8629. I think this is ready!

**Comment by lilleyse** [comment]
I spot checked sandcastle and only hit one bug: point cloud eye dome lighting crashes if log depth is false: [Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=rVVtT9swEP4rp+4DrdQmUEBiJVSDIm2TQEyDsS+VJie+NhaOHdmXljKx375L0pcUtEm8WKri3Nvz3ONzOhMOZgrn6OAEDM5hhF4VWXBb2do7SfU6soaEMuh2uvB7bIAXoXNs+ubsTEl0g1Vi4lAQ/rROy5s6pN0Zm8fO8diMTY0U+AQNBtpOhVOUZio5x5zSs2IyqWhMhPZYxYchnMKfvV2IldbKGsitMgT753CjNHoksBOgFCFRtCj3l0yUCegujIQRUkAiciocSlAG+rt7hzBnRBDg0FtdUFmT0/q7kGRBVVSCMBJS64m38WLZFnBgwPxZLVoin8BWN7lTmSI1Qx8IKdu1SA1B68f++ZJ4eyljuQqn1/J9teY7UytcgsHE2ezUc/BX2e4ffTw47HQ3SZUSI20LeZ0Kqcx0AI2S5RJEaApR9jgAcgV2t/2ZuFdZkZ02w/qbkMd6+8int3V4D9ZmN7a9lKFz3Oq2Ik8LjcM64ZPKcuuo7KodBCFhlmueCB/GRXKHFCTel/XK0A9krY5Z0wbzWCR3U2cLIwfgprFoH/S7sPrtBker3EoDVrrq/CC/b5hj63gie45lKXzT+fgEVpm8oCb4DB2pROie0GrKcmRKSo3PEXtkcxZrC3Xlii2RzZrep7BBioIJNoEnPLe9OappSgPmr+UmNQpX6kZSzUDJk3Hrya0ctyDRwnv2TAqtr9UDjlvDKOT4rTRtq0m54ja1WJQh6d7wojYGQRCF/Po8a0mbw2tOEYmY+UQUW7lY2xwb5JpG3SGnjEqDmrCo1V27WeToo5Ak43DKJnkjBZcZfq4GoIrbdkT1kRGXKWVIMbmL7T33LwWJXqyMXJmR56eeo0qKdaX/4l7YOdziFKli+2Z8bee/ZutyL+BxibL86Lwjlayq+Do2X3gs35NLyvVex+SsULocVv9mDvGq0gvQr/ifxoU/TLKcaXz7hNqy5L8phPUd42d15cqPQX07/wI)

**Comment by lilleyse** [comment]
The code looks good to me!

File: Source/Renderer/UniformState.js [hunk 1]
```diff
@@ -637,13 +641,50 @@ import SunLight from '../Scene/SunLight.js';
...
+        },
+
         /**
          * The log2 of the current frustum's far distance. Used to compute the log depth.
          * @memberof UniformState.prototype
          * @type {Number}
          */
         log2FarDistance : {
             get : function() {
+                deprecationWarning('log2FarDistance', '(czm_)log2FarDistance will be removed in Cesium 1.69.');
```
[the remark this answers is not in the document]
Comment (kring) [reply Source/Renderer/UniformState.js]: Ah I was considering the automatic uniforms to be part of the public API, but I guess they're really not. I'll just remove these entirely, then.

[the remark this answers is not in the document]
Comment (kring) [reply Source/Renderer/UniformState.js]: Ah I was considering the automatic uniforms to be part of the public API, but I guess they're really not. I'll just remove these entirely, then.

**Comment by kring** [comment]
Point cloud EDL without log depth is fixed, and those unnecessarily-deprecated uniforms have been removed.

**Comment by lilleyse** [comment]
I pushed an eslint fix: https://github.com/CesiumGS/cesium/pull/8600/commits/e979008977ca2e6b1b6fd9e1a10b05052f79204a

Will merge after CI passes

**Comment by kring** [comment]
Oops, thank you!

**Pull request merged**